File: Feature Rich Onebox For Emails/src/services/aiService.ts
import { Configuration, OpenAIApi } from 'openai';
import { config } from '../config/config';
import { Email } from '../models/email';
import { logger } from '../utils/logger';

// Simple in-memory vector store for RAG
interface VectorEntry {
  text: string;
  embedding: number[];
}

export class AiService {
  private openai: OpenAIApi;
  private apiKey: string;
  private vectorStore: VectorEntry[] = [];
  private model: string;
  private initialized: boolean = false;

  constructor() {
    this.apiKey = config.openai.apiKey;
    this.model = config.openai.model;
    
    if (this.apiKey) {
      const configuration = new Configuration({
        apiKey: this.apiKey
      });
      
      this.openai = new OpenAIApi(configuration);
      this.initialized = true;
      logger.info('AI service initialized with OpenAI');
    } else {
      logger.warn('OpenAI API key not provided. AI services will not work.');
    }
  }

  // Categorize an email
  async categorizeEmail(email: Email): Promise<string> {
    if (!this.initialized) {
      logger.warn('AI service not initialized. Returning default category.');
      return 'not_interested';
    }

    try {
      // Extract relevant information from email
      const { subject, from, text } = email;
      const truncatedText = text?.substring(0, 2000) || ''; // Limit text length for token limits
      
      const prompt = `
        Please categorize the following email into one of these categories:
        - interested (the sender shows clear interest in our product/service)
        - not_interested (neutral or negative response)
        - meeting_booked (email contains meeting confirmation)
        - spam (unsolicited email)
        - out_of_office (automatic out of office reply)
        
        Email:
        From: ${from}
        Subject: ${subject}
        
        Body:
        ${truncatedText}
        
        Category:
      `;
      
      const response = await this.openai.createChatCompletion({
        model: this.model,
        messages: [
          { role: 'system', content: 'You are an email categorization assistant. Categorize the email accurately based on its content.' },
          { role: 'user', content: prompt }
        ],
        temperature: 0.3, // Lower temperature for more deterministic results
        max_tokens: 50 // We just need a single category
      });
      
      const category = response.data.choices[0]?.message?.content?.trim().toLowerCase() || 'not_interested';
      
      // Normalize category
      const categories = ['interested', 'not_interested', 'meeting_booked', 'spam', 'out_of_office'];
      const normalizedCategory = categories.find(c => category.includes(c)) || 'not_interested';
      
      logger.info(`Categorized email "${email.subject}" as "${normalizedCategory}"`);
      return normalizedCategory;
    } catch (error) {
      logger.error(`Error categorizing email ${email.id}:`, error);
      return 'not_interested'; // Default category on error
    }
  }

  // Store text with embeddings for RAG
  async storeTextForRAG(text: string, description: string = ''): Promise<boolean> {
    if (!this.initialized) {
      logger.warn('AI service not initialized. Cannot store text for RAG.');
      return false;
    }

    try {
      const response = await this.openai.createEmbedding({
        model: 'text-embedding-ada-002',
        input: text
      });
      
      const embedding = response.data.data[0].embedding;
      
      this.vectorStore.push({
        text: text,
        embedding
      });
      
      logger.info(`Stored text for RAG: ${description || text.substring(0, 50)}...`);
      return true;
    } catch (error) {
      logger.error('Error storing text for RAG:', error);
      return false;
    }
  }

  // Find similar texts from vector store
  private async findSimilarTexts(query: string, limit: number = 3): Promise<string[]> {
    if (!this.initialized || this.vectorStore.length === 0) {
      return [];
    }

    try {
      // Get query embedding
      const response = await this.openai.createEmbedding({
        model: 'text-embedding-ada-002',
        input: query
      });
      
      const queryEmbedding = response.data.data[0].embedding;
      
      // Calculate cosine similarity with all stored embeddings
      const similarities = this.vectorStore.map(entry => {
        const similarity = this.cosineSimilarity(queryEmbedding, entry.embedding);
        return { text: entry.text, similarity };
      });
      
      // Sort by similarity (descending) and take top results
      similarities.sort((a, b) => b.similarity - a.similarity);
      const topResults = similarities.slice(0, limit);
      
      return topResults.map(result => result.text);
    } catch (error) {
      logger.error('Error finding similar texts:', error);
      return [];
    }
  }

  // Helper for cosine similarity calculation
  private cosineSimilarity(a: number[], b: number[]): number {
    let dotProduct = 0;
    let normA = 0;
    let normB = 0;
    
    for (let i = 0; i < a.length; i++) {
      dotProduct += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }
    
    return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
  }

  // Generate reply suggestion using RAG
  async generateReplySuggestion(email: Email, context: any = {}): Promise<string> {
    if (!this.initialized) {
      logger.warn('AI service not initialized. Cannot generate reply suggestion.');
      return 'AI service not available. Please compose your own reply.';
    }

    try {
      // Extract email information
      const { subject, from, text } = email;
      const truncatedText = text?.substring(0, 1000) || '';
      
      // Find relevant context from vector store
      const query = `${subject} ${truncatedText.substring(0, 200)}`;
      const similarTexts = await this.findSimilarTexts(query);
      
      const contextText = similarTexts.length > 0 
        ? `Relevant context:\n${similarTexts.join('\n\n')}`
        : 'No specific context available.';
      
      // Additional context from params
      const additionalContext = context.description || '';
      
      const prompt = `
        Please suggest a professional and helpful reply to the following email:
        
        From: ${from}
        Subject: ${subject}
        
        Email content:
        ${truncatedText}
        
        ${contextText}
        
        ${additionalContext ? `Additional context: ${additionalContext}` : ''}
        
        Reply:
      `;
      
      const response = await this.openai.createChatCompletion({
        model: this.model,
        messages: [
          { role: 'system', content: 'You are an email assistant. Generate a professional, concise, and helpful email reply.' },
          { role: 'user', content: prompt }
        ],
        temperature: 0.7,
        max_tokens: 500
      });
      
      const suggestion = response.data.choices[0]?.message?.content?.trim() || 'Unable to generate a reply.';
      
      logger.info(`Generated reply suggestion for email "${subject}"`);
      return suggestion;
    } catch (error) {
      logger.error(`Error generating reply suggestion for email ${email.id}:`, error);
      return 'Error generating reply. Please compose your own reply.';
    }
  }
}

File: Feature Rich Onebox For Emails/services/ai_service.py
import json
import logging
import os
import numpy as np
from openai import OpenAI
from models import VectorEntry, Email
from app import db

logger = logging.getLogger(__name__)

class AiService:
    """Service for AI-powered email categorization, RAG, and reply suggestions"""
    
    def __init__(self):
        self.client = None
        self.api_key = os.environ.get('OPENAI_API_KEY')
        self.model = os.environ.get('OPENAI_MODEL', 'gpt-3.5-turbo')
        self.embedding_model = os.environ.get('OPENAI_EMBEDDING_MODEL', 'text-embedding-3-small')
        self.initialized = False
        
    def initialize(self):
        """Initialize OpenAI client"""
        if self.initialized:
            return True
            
        try:
            if not self.api_key:
                logger.warning("OpenAI API key not found in environment")
                return False
                
            self.client = OpenAI(api_key=self.api_key)
            self.initialized = True
            logger.info("OpenAI client initialized")
            return True
        except Exception as e:
            logger.error(f"OpenAI initialization error: {str(e)}")
            return False
    
    def categorize_email(self, email):
        """Categorize an email using OpenAI
        
        Categories:
        - interested: Shows genuine interest in the product/service
        - not_interested: Clearly not interested
        - meeting_booked: Has booked or wants to book a meeting
        - spam: Unsolicited or irrelevant
        - out_of_office: Automated out of office reply
        """
        if not self.initialized:
            self.initialize()
            
        if not self.initialized:
            logger.error("Cannot categorize email: OpenAI not initialized")
            return "uncategorized"
            
        try:
            # Prepare email content for analysis
            email_content = f"""
            Subject: {email.subject}
            From: {email.sender}
            
            {email.body_text or ''}
            """
            
            # Create prompt for categorization
            prompt = f"""
            Categorize the following email into exactly one of these categories:
            - interested: Shows genuine interest in the product/service
            - not_interested: Clearly not interested
            - meeting_booked: Has booked or wants to book a meeting
            - spam: Unsolicited or irrelevant
            - out_of_office: Automated out of office reply
            
            Email:
            {email_content}
            
            Category (return only the category name):
            """
            
            # Call OpenAI API
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": "You are an email categorization assistant. Categorize the email into exactly one category."},
                    {"role": "user", "content": prompt}
                ],
                temperature=0.0,  # Use deterministic output
                max_tokens=20     # We only need a short response
            )
            
            # Extract category from response
            category = response.choices[0].message.content.strip().lower()
            
            # Validate category
            valid_categories = ['interested', 'not_interested', 'meeting_booked', 'spam', 'out_of_office']
            if category not in valid_categories:
                logger.warning(f"Invalid category returned: {category}")
                # Default to uncategorized if the AI returns an invalid category
                return "uncategorized"
                
            logger.info(f"Categorized email {email.id} as '{category}'")
            return category
            
        except Exception as e:
            logger.error(f"Error categorizing email: {str(e)}")
            return "uncategorized"
    
    def generate_reply_suggestion(self, email):
        """Generate an AI-powered reply suggestion for an email"""
        if not self.initialized:
            self.initialize()
            
        if not self.initialized:
            logger.error("Cannot generate reply: OpenAI not initialized")
            return "Unable to generate reply suggestion. OpenAI API not available."
            
        try:
            # Get email context
            email_content = f"""
            Subject: {email.subject}
            From: {email.sender}
            
            {email.body_text or ''}
            """
            
            # Get relevant context from vector store if available
            context_texts = self._find_similar_texts(email.body_text or '', limit=3)
            context = "\n\n".join(context_texts) if context_texts else ""
            
            # Create prompt for reply generation
            prompt = f"""
            Here is an email I've received:
            
            {email_content}
            
            """
            
            if context:
                prompt += f"""
                Here is some additional context that might be relevant:
                
                {context}
                
                """
                
            prompt += """
            Please draft a professional, helpful, and concise reply to this email.
            """
            
            # Call OpenAI API
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": "You are an email assistant. Draft professional, helpful, and concise replies."},
                    {"role": "user", "content": prompt}
                ],
                temperature=0.7,  # Some creativity
                max_tokens=500    # Reasonably sized email
            )
            
            # Extract reply from response
            reply = response.choices[0].message.content.strip()
            
            return reply
            
        except Exception as e:
            logger.error(f"Error generating reply: {str(e)}")
            return "Unable to generate reply suggestion. Error occurred."
    
    def store_text_for_rag(self, text, description=None):
        """Store text in vector database for RAG"""
        if not self.initialized:
            self.initialize()
            
        if not self.initialized:
            logger.error("Cannot store text: OpenAI not initialized")
            return False
            
        try:
            # Generate embedding for text
            embedding = self._get_embedding(text)
            
            if not embedding:
                return False
                
            # Store in database
            vector_entry = VectorEntry(
                text=text,
                embedding=json.dumps(embedding),
                description=description
            )
            
            db.session.add(vector_entry)
            db.session.commit()
            
            logger.info(f"Stored text in vector database: {description or 'Unknown'}")
            return True
            
        except Exception as e:
            logger.error(f"Error storing text: {str(e)}")
            db.session.rollback()
            return False
    
    def _get_embedding(self, text):
        """Get embedding vector for text"""
        try:
            response = self.client.embeddings.create(
                model=self.embedding_model,
                input=text
            )
            
            return response.data[0].embedding
            
        except Exception as e:
            logger.error(f"Error getting embedding: {str(e)}")
            return None
    
    def _find_similar_texts(self, query, limit=3):
        """Find similar texts in vector database"""
        if not self.initialized:
            self.initialize()
            
        if not self.initialized:
            return []
            
        try:
            # Get embedding for query
            query_embedding = self._get_embedding(query)
            
            if not query_embedding:
                return []
                
            # Get all vectors from database
            vector_entries = VectorEntry.query.all()
            
            if not vector_entries:
                return []
                
            # Calculate similarity
            results = []
            for entry in vector_entries:
                # Parse stored embedding
                entry_embedding = json.loads(entry.embedding)
                
                # Calculate cosine similarity
                similarity = self._cosine_similarity(query_embedding, entry_embedding)
                
                results.append((entry, similarity))
                
            # Sort by similarity (highest first)
            results.sort(key=lambda x: x[1], reverse=True)
            
            # Return most similar texts
            return [entry.text for entry, _ in results[:limit]]
            
        except Exception as e:
            logger.error(f"Error finding similar texts: {str(e)}")
            return []
    
    def _cosine_similarity(self, vec1, vec2):
        """Calculate cosine similarity between two vectors"""
        try:
            vec1 = np.array(vec1)
            vec2 = np.array(vec2)
            
            dot_product = np.dot(vec1, vec2)
            norm1 = np.linalg.norm(vec1)
            norm2 = np.linalg.norm(vec2)
            
            return dot_product / (norm1 * norm2)
            
        except Exception as e:
            logger.error(f"Error calculating similarity: {str(e)}")
            return 0.0